README.md
# addressbook

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar addressbook-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-07-21
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-07-21
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/addressbook/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/addressbook/compare/0.1.0...0.1.1


test/addressbook/core_test.clj
 (ns addressbook.core-test
  (:require [clojure.test :refer :all]
            [addressbook.core :refer :all]
            [ring.mock.request :as mock]))




(deftest get-values-test
  (is (= (get-values :id (mock/request :get "2"))
         {:status  404
          :headers {"content-type" "text/plain"}
          :body    "Requested record not found!"})))

;;id is a string not a number

(deftest enter-data-test
  (let [resp (enter-data (mock/request :post "/address/"  "{ \"name\" : \"rub\",
 \"email\" : \"rub@helpshift.com\",
 \"phone\" : \"1656033225\",
 \"address\" : {
   \"house\" : \"420\",
   \"apartment\" : \"Bellview Complex\",
   \"city\" : \"San Francisco\",
   \"state\" : \"CA\",
   \"zip\" : \"94107\"
 }
}"))]
    (def response resp)
    (is (= 201
           (:status resp)))
    (is (= {"content-type" "text/plain"}
           (:headers resp)))
    (is (.startsWith (:body resp)
                     "New record entered with ID: " ))))


(deftest delete-by-id-test
  (is (= (delete-by-id (mock/request :delete "2"))
         {:status  404
          :headers {"content-type" "text/plain"}
          :body    "Requested record does not exist!"}))
  )








src/addressbook/core.clj
(ns addressbook.core
  (:require [ring.adapter.jetty :as ring]
            [cheshire.core :as json]
            [schema.core :as s :include-macros true])
  (:use compojure.core)
  (:import java.util.UUID))


(defonce BOOK (atom []))  


(def email-regex #".+\@.+\..+")
(def phn-regex #"\d\d\d\d\d\d\d\d\d\d")
(def zip-regex #"\d\d\d\d\d")


(def data-schema
 {(s/required-key :name) s/Str
  (s/required-key :email) email-regex
  (s/optional-key :phone) phn-regex
  (s/optional-key :address) {(s/required-key :house) s/Str
                             (s/required-key :apartment) s/Str
                             (s/required-key :city) s/Str
                             (s/required-key :state) s/Str
                             (s/required-key :zip) zip-regex}})


(defn process-body
  "Accepts the body of request as an argument and converts the input 
  request from JSON into a clojure map and returns this map."
  [request]
  (s/validate data-schema (json/parse-string (slurp (:body request)) true)))


(defn gen-id
  "Returns a unique id."
  []
  (str (UUID/randomUUID)))

(defn get-values-as-map
  "Returns the record if the value argument exists for the key"
  [k v]
  (let [a (some (fn [x] (if (= (.toLowerCase v) (.toLowerCase (get x k))) x ))@BOOK)]a))


(defn exists?
  "Checks if the key value pair already exists in the addressbook"
  [k w]
  (if (get-values-as-map k w )
    true
    false))


(defn enter-data
  "Takes a map as an argument and adds a unique ID to it and enters it 
  into the addressbook and returns the ID of the record."
  [request]
  (let [id (gen-id)
        body (merge {:id id} (process-body request))]
    (if (or (exists? :name (get body :name)) (exists? :email (get body :email)))
      {:status 412
       :headers {"content-type" "text/plain"}
       :body (str "Value exists")}
      (do (swap! BOOK conj body)
          {:status 201
           :headers {"content-type" "text/plain"}
           :body (str "New record entered with ID: " id)}))))



(defn get-by-id
  "Returns a record map matching the id."
  [id]
  (let [a (some (fn [x] (if (= id (get x :id)) x)) @BOOK)] a))


(defn get-index
  "Returns the index of the record in the addressbook based on the id in the argument."
  [id]
  (.indexOf @BOOK (get-by-id id)) )


(defn display-records
  "Displays all the records present in the address book"
  [request]
  {:status 200
   :headers {"content-type" "application/json"}
   :body (json/generate-string {:all @BOOK})})


(defn get-values
  "Display values of a record based on the key and value provided. 
  If the value is a name then the first name and last name must be separted by %20"
  [k v]
  (let [a (some (fn [x] (if (= (.toLowerCase v) (.toLowerCase (get x k))) x)) @BOOK)]
    (if a
      {:status 200
       :headers {"content-type " "application/json"}
       :body (json/generate-string a {:pretty true})}
      {:status 404
       :headers {"content-type" "text/plain"}
       :body (str "Requested record not found!")})))


(defn process-request-handler
  "Returns the body of the request in the form of a map."
  [request]
  (process-body request))


(defn update-by-id
  "Updates the record represented by the id provided as an argument."
  [handler id]
  (fn [request]
    (if (= -1 (get-index id))
      {:status 404
       :headers {"content-type" "text/plain"}
       :body (str "No record with ID " id " found!")}
      (do
        (let [response (merge {:id id} (process-request-handler request))]
          (if (or (exists? :name (get response  :name)) (exists? :email (get response :email)))
            {:status 412
             :headers {"content-type" "text/plain"}
             :body (str "Value exists")}
            {:status 200
             :headers {"content-type" "text/plain"}
             :body (do (reset! BOOK (assoc @BOOK (get-index id) response ))
                       (str "ID: " id " updated successfully!"))}))))))


(defn delete-by-id
  "Deletes the record represented by the id provided as an argument."
  [id]
  (let [index (get-index id)]
    (if (not  (= -1 (get-index id)))
      (do (reset! BOOK (vec (concat (subvec @BOOK 0 index) (subvec @BOOK (inc index)))))
          {:status 200
           :headers {"content-type " "text/plain"}
           :body (str "ID: " id " deleted!")})
      {:status 404
       :headers {"content-type" "text/plain"}
       :body (str "Requested record does not exist!")})))


(defroutes address-book
  (GET "/address" [] display-records)  
  (GET "/address/:id" [id] (get-values :id id))      
  (GET "/address/search/:name" [name] (get-values :name name))
  (POST "/address" [] enter-data)           
  (PUT "/address/:id" [id] (update-by-id process-request-handler id))
  (DELETE "/address/:id" [id] (delete-by-id id)))  


